lib/mongodb/native_mongodb.nit

```diff
@@ -0,0 +1,501 @@
+
+	# Utility method to set `Sys.last_mongoc_error`.
+	fun set_mongoc_error(err: BSONError) do sys.last_mongoc_error = err
+end
+
+# Wrapper for `bson_error_t`.
+#
+# The `bson_error_t` structure is used to encapsulate information about an error.
+#
+# See [`bson_error_t`](http://api.mongodb.org/libbson/current/bson_error_t.html).
+extern class BSONError `{ bson_error_t * `}
+
+	# Wrapper for `error.domain`.
+	#
+	# The `error.domain` field contains the logical domain within a library that
+	# created the error.
+	fun domain: Int `{ return recv->domain; `}
+
+	# Wrapper for `error.code`.
+	#
+	# The `error.code` field contains the domain specific error code.
+	fun code: Int `{ return recv->code; `}
+
+	# Wrapper for `error.message`.
+	#
+	# The `error.message` field contains a human printable error message.
+	fun message: NativeString `{ return recv->message; `}
+end
+
+redef class Sys
+	# Last error raised by `monogdb::MongoClient`.
+	#
+	# See `MongoClient::last_error`.
+	var last_mongoc_error: nullable BSONError = null
+end
+
+# Wrapper for `mongoc_client_t`.
+#
+# `mongoc_client_t` is an opaque type that provides access to a MongoDB node,
+# replica-set, or sharded-cluster.
+# It maintains management of underlying sockets and routing to individual nodes.
+#
+# See [`mongoc_client_t`](http://api.mongodb.org/c/current/mongoc_client_t.html).
+extern class NativeMongoClient `{ mongoc_client_t * `}
+
+	# Wrapper for `mongoc_client_new()`.
+	#
+	# Creates a new `mongoc_client_t` using the `uri` string provided.
+	new(uri: NativeString) import String.to_cstring `{
+		mongoc_init();
+		return mongoc_client_new(uri);
+	`}
+
+	# Wrapper for `mongoc_client_get_server_status()`.
+	#
+	# Queries the server for the current server status.
+	# Returns `null` if an error occured.
+	fun server_status: nullable NativeBSON import set_mongoc_error, NativeBSON.as nullable `{
+		bson_error_t error;
+		bson_t reply;
+		if(!mongoc_client_get_server_status(recv, NULL, &reply, &error)){
+			NativeMongoClient_set_mongoc_error(recv, &error);
+			return null_NativeBSON();
+		}
+		//FIXME how to free the reply allocated by mongoc?
+		return NativeBSON_as_nullable(&reply);
+	`}
+
+	# Wrapper for `mongoc_client_get_database_names()`.
+	#
+	# This function queries the MongoDB server for a list of known databases.
+	# Returns `null` if an error occured.
+	fun database_names: nullable Array[NativeString]
+		import set_mongoc_error, Array[NativeString], Array[NativeString].add,
+		Array[NativeString].as nullable `{
+		Array_of_NativeString res = new_Array_of_NativeString();
+		bson_error_t error;
+		char **strv;
+		unsigned i;
+		if(strv = mongoc_client_get_database_names(recv, &error)) {
+			for(i = 0; strv[i]; i++) {
+				char *str = strv[i];
+				Array_of_NativeString_add(res, str);
+			}
+			free(strv);
+		} else {
+			free(strv);
+			NativeMongoClient_set_mongoc_error(recv, &error);
+			return null_Array_of_NativeString();
+		}
+		return res;
```

Comment:
It would have been easier to write this in Nit. By creating a private extern class for the return of `mongoc_client_database_names` which is guess is `char**`, it would only need a `fun []: NativeString`. The you do the iterating part in Nit, and use `is_address_null` to check for C NULLs.